samcli/local/apigw/local_apigw_service.py

```diff
@@ -321,11 +319,46 @@ def _query_string_params(flask_request):
             # if the list is empty, default to empty string
             if not query_string_value_length:
                 query_string_dict[query_string_key] = ""
+                multi_value_query_string_dict[query_string_key] = [""]
             else:
-                # APIGW doesn't handle duplicate query string keys, picking the last one in the list
                 query_string_dict[query_string_key] = query_string_list[-1]
+                multi_value_query_string_dict[query_string_key] = query_string_list
 
-        return query_string_dict
+        return query_string_dict, multi_value_query_string_dict
+
+    @staticmethod
+    def _event_headers(flask_request, port):
+        """
+        Constructs an APIGW equivalent headers dictionary
+
+        Parameters
+        ----------
+        flask_request request
+            Request from Flask
+        int port
+            Forwarded Port
+
+        Returns dict (str: str), dict (str: list of str)
+        -------
+            Returns a dictionary of key to list of strings
+
+        """
+        headers_dict = {}
+        multi_value_headers_dict = {}
+
+        # Multi-value request headers is not really supported by Flask.
+        # See https://github.com/pallets/flask/issues/850
+        for header_key in flask_request.headers.keys():
```

Comment:
What do you think about logging some information about not fully supporting multiple headers? We would have to print this each time, since we have no idea if the customer is impacted but would result in two things:

1) It would give the customer some information when invoking (ideally linking out to a Github issue). This could result in less of a "Wait is this me or is this SAM CLI not giving the function the headers". 
2) It would give us (SAM CLI) a way to get feedback on how many customers are impacted by this through the Github Issue +1s. 

I know this is current state and I think the only alternative is to rewrite the local service or possibly patch Flask in our implementation so we can support this. I just want to try and avoid the confusion and frustration that may come about from this.

Reply:
Good idea, what message should be displayed? Agree, it's time to consider patching Flask or rewriting the local service.

Reply:
Maybe something like:
"WARNING: Multi-value request headers are not fully supported. See issue: <link to sam cli github issue> for more details."

We would just need to create the issue with some details on how Flask doesn't support this and therefore we currently can't.

Thoughts?